lib/uri_checker.rb

```diff
@@ -0,0 +1,213 @@
+
+      report
+    end
+
+  private
+
+    def check_redirects
+      report.errors[:too_many_redirects] << "Too many redirects." if redirect_history.length >= REDIRECT_LIMIT
+      report.errors[:cyclic_redirects] << "Has a cyclic redirect." if redirect_history.include?(uri)
+      report.warnings[:multiple_redirects] << "Multiple redirects." if redirect_history.length == REDIRECT_WARNING
+    end
+
+    def check_top_level_domain
+      tld = uri.host.split(".").last
+      if INVALID_TOP_LEVEL_DOMAINS.include?(tld)
+        report.warnings[:risky_tld] << "Potentially suspicious top level domain."
+      end
+    end
+
+    def check_credentials
+      if uri.user.present? || uri.password.present?
+        report.warnings[:credentials_in_uri] << "Credentials in URI"
+      end
+    end
+
+    def check_head_request
+      start_time = Time.now
+      response = make_request(:head)
+      end_time = Time.now
+      response_time = end_time - start_time
+
+      report.warnings[:slow_response] << "Slow response time" if response_time > SLOW_RESPONSE_TIME
+
+      return response if report.has_errors?
+
+      if response.status >= 400 && response.status < 500
+        report.errors[:http_client_error] << "Received 4xx response"
+      elsif response.status >= 500 && response.status < 600
+        report.errors[:http_server_error] << "Received 5xx response"
+      end
+
+      report.warnings[:http_non_200] << "Page not available." unless response.status == 200
+      response
+    end
+
+    def check_get_request
+      response = make_request(:get)
+      return unless response
+
+      page = Nokogiri::HTML(response.body)
+      rating = page.css("meta[name=rating]")[0]["value"]
+      if %w(restricted mature).include?(rating)
+        report.warnings[:meta_rating] << "Page suggests it contains mature content."
+      end
+    end
+
+    def check_google_safebrowsing
+      api_key = Rails.application.secrets.google_api_key
+
+      response = Faraday.post do |req|
+        req.url "https://safebrowsing.googleapis.com/v4/threatMatches:find?key=#{api_key}"
+        req.headers["Content-Type"] = "application/json"
+        req.body = {
+          client: {
+            clientId: "gds-link-checker", clientVersion: "0.1.0"
+          },
+          threatInfo: {
+            threatTypes: %w(THREAT_TYPE_UNSPECIFIED MALWARE SOCIAL_ENGINEERING UNWANTED_SOFTWARE POTENTIALLY_HARMFUL_APPLICATION),
+            platformTypes: %w(ANY_PLATFORM),
+            threatEntryTypes: %w(URL),
+            threatEntries: [{ url: uri.to_s }]
+          }
+        }.to_json
+      end
+
+      data = JSON.parse(response.body)
+      if data.include?("matches") && data["matches"]
+        report.warnings[:google_safebrowsing] << "Google Safebrowsing has detected a threat."
+      end
+    end
+
+    def make_request(method)
+      begin
+        response = run_connection_request(method)
+
+        if response.status >= 300 && response.status < 400
```

Comment:
We might want to limit the range of redirects we check or check for presence of location header. I know there's some weird ones: http://www.restapitutorial.com/httpstatuscodes.html and then there's also ones we shouldn't get like 304 but would cause us pain if we did get them.